graylog2-server/src/main/java/org/graylog2/events/ClusterEventService.java

```diff
@@ -0,0 +1,203 @@
+        final BasicDBList and = new BasicDBList();
+        and.add(producerClause);
+        and.add(consumersClause);
+        final DBObject query = new BasicDBObject("$and", and);
+
+        return dbCollection.find(query)
+                .sort(DBSort.asc("$natural"))
+                .addOption(Bytes.QUERYOPTION_TAILABLE)
+                .addOption(Bytes.QUERYOPTION_AWAITDATA);
+    }
+
+    private void updateConsumers(final String eventId, final NodeId nodeId) {
+        final WriteResult<ClusterEvent, String> writeResult = dbCollection.updateById(eventId, DBUpdate.addToSet("consumers", nodeId.toString()));
+    }
+
+    private Object extractPayload(Object payload, String eventClass) {
+        try {
+            final Class<?> clazz = Class.forName(eventClass);
+            return objectMapper.convertValue(payload, clazz);
+        } catch (ClassNotFoundException e) {
+            LOG.debug("Couldn't load class <" + eventClass + "> for event", e);
+            return null;
+        } catch (IllegalArgumentException e) {
+            LOG.debug("Error while deserializing payload", e);
+            return null;
+
+        }
+    }
+
+    @Override
+    protected void run() throws Exception {
+        while (isRunning()) {
+            try {
+                LOG.debug("Opening MongoDB cursor on \"{}\"", COLLECTION_NAME);
+                final DBCursor<ClusterEvent> cursor = eventCursor(nodeId);
+                while (cursor.hasNext()) {
+                    ClusterEvent clusterEvent = cursor.next();
+                    LOG.trace("Processing cluster event: {}", clusterEvent);
+
+                    Object payload = extractPayload(clusterEvent.payload(), clusterEvent.eventClass());
+                    if (payload != null) {
+                        serverEventBus.post(payload);
+                    } else {
+                        LOG.warn("Couldn't extract payload of cluster event with ID <{}>", clusterEvent.id());
+                        LOG.debug("Invalid payload in cluster event: {}", clusterEvent);
+                    }
+
+                    updateConsumers(clusterEvent.id(), nodeId);
+                }
+            } catch (Exception e) {
+                LOG.warn("Error while reading cluster events from MongoDB, retrying.", e);
+            }
+
+            // Don't overwhelm the server
+            Uninterruptibles.sleepUninterruptibly(1L, TimeUnit.SECONDS);
+        }
+    }
+
+    @Subscribe
+    public void publishClusterEvent(Object event) {
+        if(event instanceof DeadEvent) {
+            LOG.debug("Skipping DeadEvent on cluster event bus");
+            return;
+        }
+
+        final String className = event.getClass().getCanonicalName();
+        final ClusterEvent clusterEvent = ClusterEvent.create(nodeId.toString(), className, event);
+        final String id = dbCollection.save(clusterEvent).getSavedId();
```

Comment:
Can we handle the MongoException here in any sensible manner?
At least I think we should catch the exception and log it, and not rely on Guava's default behavior (they state that one should not rely on EventBus to catch the exceptions: http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/eventbus/EventBus.html)